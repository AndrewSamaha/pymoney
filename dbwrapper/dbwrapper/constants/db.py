dbfilepath = '../db/money.sqlite'
rawCsvPath = '../db/rawTransactionHistory'
connectionString = f"sqlite:///{dbfilepath}"

def getConnectionString(env='local'):
    connectionString = {
        'local': f"sqlite:///../db/money.sqlite",
        'test': 'sqlite:///tests/mock_data/mockdb.sqlite'
    }
    if env.lower() in connectionString:
        return connectionString[env.lower()]
    return connectionString['test']

def getSeedFolder(env='local'):
    seedFolder={
        'local':    '../db/seed_data/',
        'test':     './tests/mock_data/seed_data/',
        'default':  './testsmock_data/seed_data/'
    }
    if env.lower() in seedFolder:
        return seedFolder[env.lower()]
    return seedFolder['default']