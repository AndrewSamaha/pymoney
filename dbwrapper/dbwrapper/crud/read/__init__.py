from .select_accounts_df import select_accounts_df