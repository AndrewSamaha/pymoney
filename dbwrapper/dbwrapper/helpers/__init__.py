# data-pipeline.dbwrapper.helpers.__init__.py

from .dates import getDateColumns, setDateColumns
from .hash import hash, hashLength, hashType
from .query_to_df import query_to_df
from .rawQuery import rawQuery, rawQueryDf
from .sqlEngine import engine
from .summarize import summarize
from .seed import seed_worker, getSeedEngine
from .readCsv import getAccountHistory
