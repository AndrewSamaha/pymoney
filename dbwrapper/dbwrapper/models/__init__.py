# dbwrapper.models.__init__.py

from .Account import Account
from .Base import Base
from .Load import Load
from .Transaction import Transaction, TransactionFinal
