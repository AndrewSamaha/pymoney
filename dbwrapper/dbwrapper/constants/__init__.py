# data-pipeline.dbwrapper.constants.__init__.py

from .db import connectionString, getConnectionString, getSeedFolder, rawCsvPath, dbfilepath
