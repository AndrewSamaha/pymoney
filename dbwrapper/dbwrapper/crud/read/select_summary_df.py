from sqlalchemy import text
from sqlalchemy.orm import sessionmaker
import pandas as pd
from ...helpers import rawQueryDf

def select_summary_df(engine):
    QUERY = """
        SELECT
            COUNT(*) AS total_records,
            MAX(tall.postDate) AS newest_post,
            MIN(tall.postDate) AS oldest_post,
            MAX(tall.transactionDate) AS newest_transaction,
            MIN(tall.transactionDate) AS oldest_transaction
        FROM transactions_final AS tfinal
            JOIN transactions_all AS tall ON tfinal.id = tall.id;
    """
    
    df = rawQueryDf(QUERY, engine)
    return df
