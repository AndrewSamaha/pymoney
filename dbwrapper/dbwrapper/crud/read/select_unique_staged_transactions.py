from sqlalchemy.orm import sessionmaker
from ...models import Transaction

def select_unique_staged_transactions(engine):
    Session = sessionmaker(bind=engine)
    session = Session()

    query = session.query(
            Transaction.sha256,
            Transaction.postDate,
            Transaction.transactionDate,
            Transaction.accountId,
            Transaction.amount,
            Transaction.description,
            Transaction.memo,
            Transaction.details,
            Transaction.type,
            Transaction.category,
            Transaction.balance,
            Transaction.checkOrSlipNumber
        ).group_by(
            Transaction.sha256,
            Transaction.postDate,
            Transaction.transactionDate,
            Transaction.accountId,
            Transaction.amount,
            Transaction.description,
            Transaction.memo,
            Transaction.details,
            Transaction.type,
            Transaction.category,
            Transaction.balance,
            Transaction.checkOrSlipNumber,
        ).order_by(
            Transaction.postDate
        )

    session.close()
    
    return query
