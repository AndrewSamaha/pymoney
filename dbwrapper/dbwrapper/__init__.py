from .constants import connectionString

from .models.Base import Base
from .models.Account import Account
from .models.Transaction import TransactionFinal, Transaction
from .models.Load import Load

from .helpers.sqlEngine import engine
from .helpers.query_to_df import query_to_df

from .helpers.readCsv import setDateColumns
from .helpers import rawQuery, rawQueryDf

from .crud.read.select_summary_df import select_summary_df

