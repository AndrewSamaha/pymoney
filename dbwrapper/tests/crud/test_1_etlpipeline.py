import pytest
import os
import pandas as pd
from sqlalchemy.orm import sessionmaker
from dbwrapper.constants import getConnectionString, getSeedFolder
from dbwrapper.helpers import seed_worker, getSeedEngine, getAccountHistory, openOneCsv, prepCheckingDf
from dbwrapper.models import Base
from dbwrapper.crud.read import select_accounts_df

rawCsvPath='tests/mock_data/rawTransactionHistory'

class TestEtlPipeline:
    @pytest.fixture(scope='class')
    def sqlite_file(self, tmp_path_factory):
        fn = tmp_path_factory.getbasetemp() / 'mockdb.sqlite'
        return fn

    @pytest.fixture
    def connection(self, sqlite_file):
        ENV = 'Test'
        dryrun = False
        seedFolder = getSeedFolder(ENV)
        connectionString = f"sqlite:///{str(sqlite_file)}"
        engine = getSeedEngine(connectionString)
        return {
            'ENV': ENV,
            'dryrun': dryrun,
            'seedFolder': seedFolder,
            'connectionString': connectionString,
            'engine': engine
        }
    
    @pytest.fixture
    def engine(self, connection):
        return connection['engine']

    def test_create_and_seed_db(self, connection, sqlite_file, snapshot):
        seed_worker(
            connection['seedFolder'],
            connection['engine'],
            connection['dryrun'],
            Base.metadata)
        self.accounts = select_accounts_df(connection['engine'])
        accountString = self.accounts.to_csv()
        assert accountString == snapshot

    def test_prep_checking_df(self):
        # dbwrapper/helpers/readCsv
        input_columns = [
            'Details',
            'Description',
            'Amount',
            'Type',
            'Balance',
            'Check or Slip #',
            'Post Date',
            'Transaction Date',
            'Category'
        ]
        df = pd.DataFrame(columns=input_columns)
        result = list(prepCheckingDf(df, '0').columns)
        expected_columns = [
            'details',
            'description',
            'amount',
            'type',
            'balance',
            'checkOrSlipNumber',
            'postDate',
            'transactionDate',
            'category',
            'loadID'
        ]
        print(result)
        assert result == expected_columns
    
    def test_open_one_csv(self, connection, engine, snapshot):
        # dbwrapper/helpers/readCsv
        accounts = select_accounts_df(engine)
        resultsString = ""
        df = None
        for index, account in accounts.iterrows():
            path = f"{account['csvPath']}"
            accountPath = f"{rawCsvPath}/{path}"
            for file in os.listdir(accountPath):
                fileAndPath = f"{rawCsvPath}/{path}/{file}"
                filenameParts = file.split('.')
                filename = filenameParts[:len(filenameParts)-1][0]
                df = openOneCsv(fileAndPath)
                break
            break
        assert df.to_csv() == snapshot

    def test_get_account_history(self, connection, engine, snapshot):
        accounts = select_accounts_df(engine)
        resultsString = ""
        for index, account in accounts.iterrows():
            path = f"{account['csvPath']}"
            Session = sessionmaker(bind=engine)
            session = Session()
            allFiles = getAccountHistory(path, account['id'], session, rawCsvPath=rawCsvPath)
            print('test_get_account_history')
            resultsString = f"{resultsString}{allFiles.to_csv()}"
            print(allFiles.to_csv())

        assert resultsString==snapshot
