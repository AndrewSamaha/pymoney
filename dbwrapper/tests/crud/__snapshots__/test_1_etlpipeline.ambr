# serializer version: 1
# name: TestEtlPipeline.test_create_and_seed_db
  '''
  ,id,name,accountNumber,interestRate,statementDate,paymentDueDate,csvPath
  0,1,Checking Account,111111111,0.0,,,checkingAccount/
  1,2,Chase Freedom,222222222,20.0,,,creditAccount/
  
  '''
# ---
# name: TestEtlPipeline.test_get_account_history
  '''
  ,details,description,amount,type,balance,checkOrSlipNumber,sha256,stagingLevel,postDate,accountId,loadID
  0,DEBIT,POS DEBIT             PUBLIX                               FL,-44.44,MISC_DEBIT,,,bbfc65a715,allTransactions,2023-06-20,1,1
  1,DEBIT,POS DEBIT             CITGO                                FL,-12.01,MISC_DEBIT,,,5a5183fb8d,allTransactions,2023-06-19,1,1
  2,DEBIT,POS DEBIT             MODERN Karate CENTER   TAMPA         FL,-108.0,MISC_DEBIT,,,2b7ed14dd0,allTransactions,2023-06-18,1,1
  0,DEBIT,POS DEBIT             MODERN Karate CENTER   TAMPA         FL,-108.0,MISC_DEBIT,,,2b7ed14dd0,allTransactions,2023-06-18,1,2
  3,DEBIT,Payment to Credit card ending in 1111                   06/17,-700.0,ACCT_XFER,1651.89,,c200376bf5,allTransactions,2023-06-17,1,1
  1,DEBIT,Payment to Credit card ending in 1111                   06/17,-700.0,ACCT_XFER,1651.89,,c200376bf5,allTransactions,2023-06-17,1,2
  3,DEBIT,GOOGLE *Disney Mobile 855-836-3987 CA                   06/14,-12.47,DEBIT_CARD,2438.76,,dee5a7c27f,allTransactions,2023-06-14,1,2
  4,CREDIT,JOBCOMPANY  PAYROLL                       PPD ID: 0000000001,3000.0,ACH_CREDIT,10000.0,,a4099b862b,allTransactions,2023-06-14,1,2
  4,DEBIT,UTILITIES        UTIL PYMT                 PPD ID: 0000000001,-86.87,ACH_DEBIT,2351.89,,a0ffbc02fb,allTransactions,2023-06-14,1,1
  2,DEBIT,UTILITIES        UTIL PYMT                 PPD ID: 0000000001,-86.87,ACH_DEBIT,2351.89,,a0ffbc02fb,allTransactions,2023-06-14,1,2
  ,transactionDate,postDate,description,category,type,amount,Memo,sha256,stagingLevel,accountId,loadID
  0,2023-06-24,2023-06-24,REDEMPTION CREDIT,Fees & Adjustments,Adjustment,1677.19,,c05ed95317,allTransactions,2,4
  1,2023-06-23,2023-06-24,EQ *BARBAB,Shopping,Sale,-27.69,,bbedf7e175,allTransactions,2,3
  3,2023-06-23,2023-06-24,EQ *BARBAB,Shopping,Sale,-27.69,,bbedf7e175,allTransactions,2,4
  2,2023-06-23,2023-06-24,EQ *TURTLE,Food & Drink,Sale,-25.56,,8a055328d0,allTransactions,2,3
  4,2023-06-23,2023-06-24,EQ *TURTLE,Food & Drink,Sale,-25.56,,8a055328d0,allTransactions,2,4
  1,2023-06-22,2023-06-24,Coffee,Food & Drink,Sale,-9.37,,1bb122e1ee,allTransactions,2,4
  0,2023-06-22,2023-06-24,parkreceipts.com,Travel,Sale,-6.0,,1a950d3f1a,allTransactions,2,3
  2,2023-06-22,2023-06-24,parkreceipts.com,Travel,Sale,-6.0,,1a950d3f1a,allTransactions,2,4
  3,2023-06-22,2023-06-23,RRR* Hungry Socks,Food & Drink,Sale,-67.05,,601ab19ae3,allTransactions,2,3
  5,2023-06-22,2023-06-23,RRR* Hungry Socks,Food & Drink,Sale,-67.05,,601ab19ae3,allTransactions,2,4
  4,2023-06-20,2023-06-21,GUTHRIES CHICKEN,Food & Drink,Sale,-23.54,,dfba74632b,allTransactions,2,3
  6,2023-06-20,2023-06-21,GUTHRIES CHICKEN,Food & Drink,Sale,-23.54,,dfba74632b,allTransactions,2,4
  6,2023-06-20,2023-06-20,GOOGLE *YouTube,Shopping,Sale,-13.6,,9b13c82bca,allTransactions,2,3
  8,2023-06-20,2023-06-20,GOOGLE *YouTube,Shopping,Sale,-13.6,,9b13c82bca,allTransactions,2,4
  5,2023-06-19,2023-06-20,Amazon.com,Shopping,Sale,-26.88,,894221824a,allTransactions,2,3
  7,2023-06-19,2023-06-20,Amazon.com,Shopping,Sale,-26.88,,894221824a,allTransactions,2,4
  9,2023-06-18,2023-06-19,PUBLIX,Groceries,Sale,-143.53,,b48b77a378,allTransactions,2,3
  7,2023-06-18,2023-06-19,PUBLIX,Groceries,Sale,-9.49,,9bff506cdc,allTransactions,2,3
  9,2023-06-18,2023-06-19,PUBLIX,Groceries,Sale,-9.49,,9bff506cdc,allTransactions,2,4
  8,2023-06-18,2023-06-19,STARBUCKS 800-782-7282,Food & Drink,Sale,-25.0,,8c17341749,allTransactions,2,3
  
  '''
# ---
# name: TestEtlPipeline.test_open_one_csv
  '''
  ,Details,Posting Date,Description,Amount,Type,Balance,Check or Slip #,sha256,stagingLevel
  0,DEBIT,2023-06-20,POS DEBIT             PUBLIX                               FL,-44.44,MISC_DEBIT,,,bbfc65a715,allTransactions
  1,DEBIT,2023-06-19,POS DEBIT             CITGO                                FL,-12.01,MISC_DEBIT,,,5a5183fb8d,allTransactions
  2,DEBIT,2023-06-18,POS DEBIT             MODERN Karate CENTER   TAMPA         FL,-108.0,MISC_DEBIT,,,2b7ed14dd0,allTransactions
  3,DEBIT,2023-06-17,Payment to Credit card ending in 1111                   06/17,-700.0,ACCT_XFER,1651.89,,c200376bf5,allTransactions
  4,DEBIT,2023-06-14,UTILITIES        UTIL PYMT                 PPD ID: 0000000001,-86.87,ACH_DEBIT,2351.89,,a0ffbc02fb,allTransactions
  
  '''
# ---
